File: obs_replaykit/sleep_override.py
"""drop OBS power requests so Windows idle monitor/system sleep can still run."""

from __future__ import annotations

import re
import subprocess
from typing import Callable, Optional

LogFn = Optional[Callable[[str], None]]

_OBS_EXE_NAME = "obs64.exe"
_REQUEST_FLAGS = ("DISPLAY", "SYSTEM", "AWAYMODE")

# suppress the powercfg console flash when spawned from a non-console parent (pyinstaller apply exe).
_CREATE_NO_WINDOW = 0x08000000


def _run_powercfg(args: list[str], log: LogFn = None) -> Optional[subprocess.CompletedProcess]:
    """subprocess.run wrapper with our standard timeout + create_no_window flags. None on oserror/timeout."""
    try:
        return subprocess.run(
            ["powercfg.exe", *args],
            capture_output=True,
            text=True,
            timeout=10,
            creationflags=_CREATE_NO_WINDOW,
        )
    except (subprocess.TimeoutExpired, OSError) as exc:
        if log:
            log(f"warn: powercfg {' '.join(args)} failed: {exc}")
        return None


def _installed_request_flags() -> set[str]:
    result = _run_powercfg(["/requestsoverride"])
    if result is None or result.returncode != 0:
        return set()
    flags: set[str] = set()
    for line in result.stdout.splitlines():
        stripped = line.strip()
        if not stripped:
            continue
        if stripped.lower().startswith(_OBS_EXE_NAME.lower()):
            tail = stripped[len(_OBS_EXE_NAME):]
            for flag in _REQUEST_FLAGS:
                if re.search(rf"\b{re.escape(flag)}\b", tail, re.IGNORECASE):
                    flags.add(flag)
    return flags


def is_sleep_override_installed() -> bool:
    """true iff all ReplayKit-managed OBS power request overrides are already in place."""
    installed = _installed_request_flags()
    return all(flag in installed for flag in _REQUEST_FLAGS)


def install_sleep_override(log: LogFn = None) -> bool:
    """ignore OBS display/system/awaymode requests so Windows sleep timers still work."""
    if is_sleep_override_installed():
        if log:
            log(f"sleep override already in place ({_OBS_EXE_NAME} -> {'/'.join(_REQUEST_FLAGS)})")
        return True

    result = _run_powercfg(
        ["/requestsoverride", "PROCESS", _OBS_EXE_NAME, *_REQUEST_FLAGS],
        log=log,
    )
    if result is None:
        return False
    if result.returncode != 0:
        err = (result.stderr or result.stdout or "").strip().splitlines()
        err_line = err[0] if err else f"exit {result.returncode}"
        if log:
            log(f"warn: powercfg /requestsoverride failed: {err_line}")
        return False

    if log:
        log(f"sleep override installed ({_OBS_EXE_NAME} -> {'/'.join(_REQUEST_FLAGS)} requests ignored)")
        log("Windows monitor and PC sleep timers can now run while OBS/ReplayKit is active.")
    return True


def remove_sleep_override(log: LogFn = None) -> bool:
    """drop the override entry. powercfg /requestsoverride PROCESS <exe> with no request-type flags after the exe name removes the entry entirely. safe to call when no override exists (powercfg exits 0)."""
    result = _run_powercfg(
        ["/requestsoverride", "PROCESS", _OBS_EXE_NAME],
        log=log,
    )
    if result is None:
        return False
    if result.returncode != 0:
        err = (result.stderr or result.stdout or "").strip().splitlines()
        err_line = err[0] if err else f"exit {result.returncode}"
        if log:
            log(f"warn: powercfg remove override failed: {err_line}")
        return False
    if log:
        log(f"sleep override removed for {_OBS_EXE_NAME}")
    return True

File: ReplayKitSetup/SleepOverride.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text.RegularExpressions;

namespace ReplayKitSetup
{
    // drop obs power requests so windows idle monitor/system sleep can still run. ported from obs_replaykit/sleep_override.py.
    public static class SleepOverride
    {
        private const string ObsExeName = "obs64.exe";
        private static readonly string[] RequestFlags = { "DISPLAY", "SYSTEM", "AWAYMODE" };

        // subprocess.run wrapper with our standard timeout + create_no_window flags. null on win32exception/timeout.
        private static (int ExitCode, string Stdout, string Stderr)? RunPowercfg(string arguments, Action<string> log = null)
        {
            try
            {
                var psi = new ProcessStartInfo("powercfg.exe", arguments)
                {
                    UseShellExecute = false,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    CreateNoWindow = true,
                };
                using (var proc = Process.Start(psi))
                {
                    string stdout = proc.StandardOutput.ReadToEnd();
                    string stderr = proc.StandardError.ReadToEnd();
                    if (!proc.WaitForExit(10000))
                    {
                        try { proc.Kill(); } catch (InvalidOperationException) { }
                        log?.Invoke("warn: powercfg " + arguments + " failed: timed out");
                        return null;
                    }
                    return (proc.ExitCode, stdout, stderr);
                }
            }
            catch (System.ComponentModel.Win32Exception exc)
            {
                log?.Invoke("warn: powercfg " + arguments + " failed: " + exc.Message);
                return null;
            }
        }

        private static HashSet<string> InstalledRequestFlags()
        {
            var flags = new HashSet<string>();
            var result = RunPowercfg("/requestsoverride");
            if (result == null || result.Value.ExitCode != 0) return flags;
            foreach (var rawLine in result.Value.Stdout.Split('\n'))
            {
                string stripped = rawLine.Trim();
                if (stripped.Length == 0) continue;
                if (stripped.ToLowerInvariant().StartsWith(ObsExeName.ToLowerInvariant()))
                {
                    string tail = stripped.Substring(ObsExeName.Length);
                    foreach (var flag in RequestFlags)
                    {
                        if (Regex.IsMatch(tail, @"\b" + Regex.Escape(flag) + @"\b", RegexOptions.IgnoreCase)) flags.Add(flag);
                    }
                }
            }
            return flags;
        }

        // true iff all replaykit-managed obs power request overrides are already in place.
        public static bool IsSleepOverrideInstalled()
        {
            var installed = InstalledRequestFlags();
            return RequestFlags.All(installed.Contains);
        }

        // ignore obs display/system/awaymode requests so windows sleep timers still work.
        public static bool InstallSleepOverride(Action<string> log = null)
        {
            if (IsSleepOverrideInstalled())
            {
                log?.Invoke($"sleep override already in place ({ObsExeName} -> {string.Join("/", RequestFlags)})");
                return true;
            }

            var result = RunPowercfg(Win32Args.Build(new[] { "/requestsoverride", "PROCESS", ObsExeName }.Concat(RequestFlags).ToArray()), log);
            if (result == null) return false;
            if (result.Value.ExitCode != 0)
            {
                string message = !string.IsNullOrEmpty(result.Value.Stderr) ? result.Value.Stderr : result.Value.Stdout;
                string errLine = string.IsNullOrWhiteSpace(message) ? "exit " + result.Value.ExitCode : message.Trim().Split('\n')[0];
                log?.Invoke("warn: powercfg /requestsoverride failed: " + errLine);
                return false;
            }

            log?.Invoke($"sleep override installed ({ObsExeName} -> {string.Join("/", RequestFlags)} requests ignored)");
            log?.Invoke("Windows monitor and PC sleep timers can now run while OBS/ReplayKit is active.");
            return true;
        }

        // drop the override entry. powercfg /requestsoverride process <exe> with no request-type flags after the exe name removes the entry entirely. safe to call when no override exists (powercfg exits 0).
        public static bool RemoveSleepOverride(Action<string> log = null)
        {
            var result = RunPowercfg(Win32Args.Build("/requestsoverride", "PROCESS", ObsExeName), log);
            if (result == null) return false;
            if (result.Value.ExitCode != 0)
            {
                string message = !string.IsNullOrEmpty(result.Value.Stderr) ? result.Value.Stderr : result.Value.Stdout;
                string errLine = string.IsNullOrWhiteSpace(message) ? "exit " + result.Value.ExitCode : message.Trim().Split('\n')[0];
                log?.Invoke("warn: powercfg remove override failed: " + errLine);
                return false;
            }
            log?.Invoke("sleep override removed for " + ObsExeName);
            return true;
        }
    }
}
